### ISSUE TITLE
Ask for sudo/admin on files saving if required

### ISSUE THREAD

**Issue by itsbalamurali** [zed-industries/zed#13259]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Screenshot of VS Code
<img width="557" alt="image" src="https://github.com/zed-industries/zed/assets/1926303/6307ad16-e24a-4150-b179-4a521531d98e">

try editing `/etc/hosts` file in zed.

### Environment

MacOS

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="557" alt="image" src="https://github.com/zed-industries/zed/assets/1926303/6307ad16-e24a-4150-b179-4a521531d98e">

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by meduk0** [comment]
i think the best way to do it is to save a temporary copy of the modified file in a $USER repo like .cache or /tmp folder  and then move it to the original file location while using pkexec . i am working on a script for this method . any help is appreciated !

**Comment by Esteban-Rocha** [comment]
Heya Zed team!

Don't intent to add noise to the convo, rather just comment out this is still a pretty needed thing and I'll argue this is a bit priority, for additional context SublimeText does this as well:

![Image](https://github.com/user-attachments/assets/f1efe7a4-68ed-43a5-9c93-6e6bc68469f5)

btw, thanks to the ZED team, the editor is getting better and better, much appreciated.

**Comment by Malix-Labs** [comment]
> [@itsbalamurali](https://github.com/itsbalamurali) i think the best way to do it is to save a temporary copy of the modified file in a $USER repo like .cache or /tmp folder and then move it to the original file location while using pkexec . i am working on a script for this method . any help is appreciated ! this was my simple implementation in neovim (it is simple and usually barely make problems )

This is essentially re-implementing `sudoedit` / `sudo -e` behavior

**Comment by ambyjkl** [comment]
> [@itsbalamurali](https://github.com/itsbalamurali) i think the best way to do it is to save a temporary copy of the modified file in a $USER repo like .cache or /tmp folder and then move it to the original file location while using pkexec . i am working on a script for this method . any help is appreciated ! this was my simple implementation in neovim (it is simple and usually barely make problems )

a way that avoids creating temporary file is piping into `pkexec tee <outfile>`

**Comment by Malix-Labs** [comment]
This is becoming my highest pain point with Zed, as I sometimes need to edit root config files and Zed seems to also be incapable of being used for `sudoedit`, so I have to revert back to using Vscode for such uses

_**edit**: forget about it, sudoedit works now, hiding myself as resolved_

**Comment by ambyjkl** [comment]
> This is becoming my highest pain point with Zed, as I sometimes need to edit root config files and Zed seems to also be incapable of being used for `sudoedit`, so I have to revert back to using Vscode for such uses
> 
> _**edit**: forget about it, sudoedit works now, hiding myself as resolved_

You should try the Micro editor, it can run in the zed terminal and has sudo support, and similar keybinds to vscode, so don't need to open vscode

**Comment by hedefalk** [comment]
I'd love to get this too. I want to use zed everywhere and I've tried setting 

     EDITOR = "zed --wait";

to use for commands like vifs, but:

```
❯ vifs
vifs: need to run as root

~
❯ sudo vifs
Password:
Error: Running Zed as root or via sudo is unsupported.
       Doing so (even once) may subtly break things for all subsequent non-root usage of Zed.
       It is untested and not recommended, don't complain when things break.
       If you wish to proceed anyways, set `ZED_ALLOW_ROOT=true` in your environment.
vifs: editing error

❯ echo $ZED_ALLOW_ROOT
true
```

**Comment by Jason5Lee** [comment]
Is it even possible to edit a file that requires sudo permission? I cannot even run zed with sudo (saying it is untested and not recommended).

**Comment by Malix-Labs** [comment]
@Jason5Lee use `sudoedit` (or any tool that achieves a similar job)

**Comment by ghost** [comment]
> Is it even possible to edit a file that requires sudo permission? I cannot even run zed with sudo (saying it is untested and not recommended).

Exact : 
I have done a symlink : 
`sudo ln -s ~/.local/bin/zed /usr/local/bin/zed`

And now when editing a file with `sudo zed myFile` I have this warning : 
```
Error: Running Zed as root or via sudo is unsupported.
       Doing so (even once) may subtly break things for all subsequent non-root usage of Zed.
       It is untested and not recommended, don't complain when things break.
       If you wish to proceed anyways, set `ZED_ALLOW_ROOT=true` in your environment.
```

Not sure what can break

**Comment by iNViTiON** [comment]
@Jason5Lee @ghost As Malix-Labs said, you can edit with `sudoedit`
e.g. `EDITOR="zed --wait" sudoedit /etc/hosts`. `sudoedit` will responsible for copy file content to temporary file, opening that file with executable we set, then copy the temp file content back to source file. In this case `--wait` will hold `sudoedit` until we close that file in Zed before copying the content back.

Running Zed as root might touch some file (e.g. config) resulting in that file permission will need root resulting in Zed will forever need root to edit or read that config.

**Comment by DarknessEoL** [comment]
> Using sudoedit feels hacky. Zed competitors like VSCode, Sublime, even KDE Kate editor show a popup to enter credentials when saving a file that needs root.

Yeah this is the implementation I honestly expected when switching from kate to zed.

**Comment by LeoMerlino** [comment]
This feature seems to have been added in the past (#22045) but is now not a thing any more?

**Comment by Damglador** [comment]
@merll002 https://github.com/zed-industries/zed/issues/13585#issuecomment-2670577037

**Comment by psudo123** [comment]
Open issue, 126 upvotes, 0 fix

**Comment by SublimePeace** [comment]
Back to VSCode I go.. apparently adding AI features nobody asked for is higher priority than the ability to save files

**Comment by LeoMerlino** [comment]
@SublimePeace 
> Back to VSCode I go.. apparently adding AI features nobody asked for is higher priority than the ability to save files

As annoying as it may be, at the end of the day Zed is open source software and you can't just get people to fix issues that they are not interested in fixing.
We need to wait for someone who cares about this issue and has enough knowledge to contribute.
